internal/webhook/managedcluster_webhook.go

```diff
@@ -298,3 +336,38 @@ func getMissingProvidersWithWrongVersions(exposed, required []hmcv1alpha1.Provid
 
 	return missing, nonSatisfying, merr
 }
+
+func isCredMatchTemplate(cred *hmcv1alpha1.Credential, template *hmcv1alpha1.ClusterTemplate) error {
+	idtyKind := cred.Spec.IdentityRef.Kind
+	infraProviders := template.Status.Providers.InfrastructureProviders
+
+	errMsg := func(idtyKind string, provider string) error {
+		return fmt.Errorf("wrong kind of the ClusterIdentity \"%s\" for provider \"%s\"", idtyKind, provider)
+	}
+
+	if len(infraProviders) == 0 {
+		return fmt.Errorf("template \"%s\" has no infrastructure providers defined", template.Name)
+	}
+
+	for _, provider := range infraProviders {
+		switch provider.Name {
+		case "aws":
+			if idtyKind != "AWSClusterStaticIdentity" &&
+				idtyKind != "AWSClusterRoleIdentity" &&
+				idtyKind != "AWSClusterControllerIdentity" {
+				return errMsg(idtyKind, provider.Name)
+			}
+		case "azure":
+			if idtyKind != "AzureClusterIdentity" {
```

Comment:
kind of nit either: what do you think if we utilize the const from [the dedicated file](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e3e3e309b07d470fa79cb3ee172debc029c4d8f2/api/v1beta1/consts.go#L195)? Either way the dependency will be introduced by the #405 PR

If we can, then I hope there are such constants in the corresponding packages (maybe skip the `aws` ones since that dependency has not yet been brought)

side note: if you do not want to escape quotes (like in `"template \"%s\"`) you can either use the graves `` `template "%s"` `` or the [dedicated verb](https://pkg.go.dev/fmt#hdr-Printing) for double-quotes strings `template %q`

Reply:
@zerospiel I know that the kind comparison by comparing strings looks kinda shitty. Unfortunately only Azure provider has these consts. VSphere has [something else](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/apis/v1beta1/vsphereclusteridentity_types.go#L66) and AWS doesn't care at all.
So I decided to go with string comparison, since it makes it uniform across all providers